File: goblin-cache/src/main/java/org/goblinframework/cache/bean/GoblinCache.java
package org.goblinframework.cache.bean;

import org.goblinframework.cache.annotation.CacheExpiration;
import org.goblinframework.cache.core.Cache;
import org.goblinframework.cache.core.CacheLocation;
import org.goblinframework.cache.core._CacheExpirationCalculatorKt;
import org.jetbrains.annotations.NotNull;

public class GoblinCache {

  public Class<?> type;
  public CacheLocation location;
  public boolean wrapper;
  public CacheExpiration.Policy expirationPolicy = CacheExpiration.Policy.FIXED;
  public int expirationValue = CacheExpiration.DEFAULT_EXPIRATION;

  public Cache cache;

  public int calculateExpiration() {
    CacheExpiration annotation = type.getAnnotation(CacheExpiration.class);
    if (annotation != null && annotation.enable()) {
      return _CacheExpirationCalculatorKt.calculateExpirationInSeconds(annotation);
    }
    return _CacheExpirationCalculatorKt.calculateExpirationInSeconds(expirationPolicy, expirationValue);
  }

  @NotNull
  public Cache cache() {
    assert cache != null;
    return cache;
  }
}

File: goblin-cache/src/main/java/org/goblinframework/cache/bean/GoblinCache.kt
package org.goblinframework.cache.bean

import org.goblinframework.cache.annotation.GoblinCacheExpiration
import org.goblinframework.cache.core.cache.Cache
import org.goblinframework.cache.core.cache.CacheLocation
import org.goblinframework.cache.util.CacheExpirationCalculator

class GoblinCache {

  var type: Class<*>? = null
  var location: CacheLocation? = null
  var wrapper = false
  var expirationPolicy = GoblinCacheExpiration.Policy.FIXED
  var expirationValue = GoblinCacheExpiration.DEFAULT_EXPIRATION
  var cache: Cache? = null

  fun calculateExpiration(): Int {
    val annotation = type!!.getAnnotation(GoblinCacheExpiration::class.java)
    return if (annotation != null && annotation.enable) {
      CacheExpirationCalculator.expirationInSeconds(annotation)
    } else CacheExpirationCalculator.expirationInSeconds(expirationPolicy, expirationValue)
  }

  fun cache(): Cache {
    return cache!!
  }
}